x/authz/count_authorization.go

```diff
@@ -0,0 +1,49 @@
+package authz
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var (
+	_ Authorization = &CountAuthorization{}
+)
+
+// NewCountAuthorization creates a new CountAuthorization object.
+func NewCountAuthorization(msgTypeURL string, allowedAuthorizations int32) *CountAuthorization {
+	return &CountAuthorization{
+		Msg:                   msgTypeURL,
+		AllowedAuthorizations: allowedAuthorizations,
+	}
+}
+
+// MsgTypeURL implements Authorization.MsgTypeURL.
+func (a CountAuthorization) MsgTypeURL() string {
+	return a.Msg
+}
+
+// Accept implements Authorization.Accept.
+func (a CountAuthorization) Accept(ctx sdk.Context, msg sdk.Msg) (AcceptResponse, error) {
+	remaining, isNegative := a.decrement()
+	if isNegative {
+		return AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf("allowed authorizations cannot be negative")
```

Comment:
This should be a `ErrUnauthorized` or `ErrorInvalidSigner` (not sure which is better here).
And a better error message would be something like "no authorized uses left". It's confusing (from the outside) that it's referencing the number after the decrement rather than what it currently is. That is, the number is actually 0 when they make the request. It's not negative.